File: rrs_unity/Assets/Scripts/RRS/Network/Net2/Main/Objects/Net2Subscirber.cs
using RRS.Tools.Log;
using RRS.Tools.Protobuf;
using System;
using System.Collections.Generic;
using System.Text;
using NetMQ;
using NetMQ.Sockets;
using NetMQ.Monitoring;
using ProtoBuf;
using System.Threading;
using System.Threading.Tasks;
using Consul;

namespace RRS.Tools.Network
{
    public partial class Net2
    {
        private class Net2Subscriber : Net2Base
        {
            public delegate ProcessResult DelegateGetServiceInfo(string name);
            public event DelegateGetServiceInfo delegateGetServiceInfo;

            public delegate void DelegateNewData(long sequence, byte[] buffer, uint priority, ulong id);
            public event DelegateNewData delegateNewData;

            AgentService info;

            public Net2Subscriber(string topic_name) : base()
            {
                this.name = topic_name;
            }

            protected ProcessResult getServiceInfo(string name)
            {
                ProcessResult result = new ProcessResult();

                if (delegateGetServiceInfo != null)
                {
                    return delegateGetServiceInfo(name);
                }
                else
                {
                    result.Success = false;
                    result.Result = null;
                    result.ResultType = null;
                }

                return result;
            }

            public AgentService GetAddress(string name)
            {
                ProcessResult result = getServiceInfo(name);

                if (result.Success && result.Result != null && result.Result is AgentService)
                {
                    return result.Result as AgentService;
                }
                else
                {
                    return null;
                }
            }

            protected override void _SocketBuilder()
            {
                if (info == null)
                {
                    info = GetAddress(name);
                    if (info == null)
                    {
                        Thread.Sleep(900);
                        return;
                    }
                }
                reportLog("Starting subscriber " + Name, LogType.INFO, section);

                socket = new SubscriberSocket();

                socket.Options.TcpKeepalive = true;
                socket.Options.TcpKeepaliveIdle = TimeSpan.FromMilliseconds(100);
                socket.Options.TcpKeepaliveInterval = TimeSpan.FromMilliseconds(100);

                string path = "tcp://" + info.Address + ":" + info.Port;

                ((SubscriberSocket)socket).Subscribe("");

                //monitor = new NetMQMonitor(socket, "inproc://sub" + Net2.GetAddressIndex(), SocketEvents.All);

                //monitor.Disconnected += Monitor_Disconnected;
                //monitor.Connected += Monitor_Connected;

                //monitor.Timeout = TimeSpan.FromMilliseconds(100);
                //task = Task.Factory.StartNew(monitor.Start);

                setState(Net2State.STARTED);
                start_time = GetTime();

                socket.Connect(path);

                reportLog("Subscriber " + Name + " is connecting to " + path, LogType.INFO, section);
            }

            protected override void _CheckState()
            {
                if (internal_state == Net2State.STARTED)
                {
                    long diff = GetTime() - start_time;
                    if (diff > 2000)
                    {
                        reportLog("Stopping cause monitor failure", LogType.WARN, section);
                        throw new Exception("monitor failure");
                    }
                }
                else
                if (internal_state == Net2State.CONNECTED)
                {
                    long diff = GetTime() - last_send_receive_time;
                    if (diff > 5000)
                    {
                        Monitor_Disconnected(null,null);
                    }
                }
            }

            protected override void _Send()
            {

            }

            protected override void threadExitProcess()
            {
               
            }

            protected override void Monitor_Disconnected(object sender, NetMQMonitorSocketEventArgs e)
            {
                reportLog("Monitor disconnected is called " + name, LogType.INFO, section);
                thread_exit = true;
                Dispose(false);
                reportCrash(); 
            }

            protected override void Monitor_Connected(object sender, NetMQMonitorSocketEventArgs e)
            {
                base.Monitor_Connected(sender, e);
                setState(Net2State.CONNECTED);
            }

            protected override void reportData(Message packet)
            {
                last_send_receive_time = GetTime();

                if (internal_state == Net2State.STARTED)
                {
                    Monitor_Connected(null, null);
                }

                delegateNewData?.Invoke(packet.header.sequence, packet.payload, packet.header.priority, Id);
            }

            protected override void StopMonitor()
            {
                if (monitor != null)
                {
                    monitor.Disconnected -= Monitor_Disconnected;
                    monitor.Connected -= Monitor_Connected;
                    monitor.Stop();
                }
            }
        }
    }
}

File: rrs_ros/src/core/src/Net2/Main/Objects/Net2Subscirber.cpp
#include "Net2/Net2Subscirber.h"
#include "Net2/Enum.h"
#include "Net2/Net2.h"

namespace lmt
{
namespace Tools
{
namespace Network
{

Net2Subscriber::Net2Subscriber(void* context) : Net2Base("", Net2NetworkType::SUBSCRIBER,context)
{

}

ProcessResult<int> Net2Subscriber::Start(const std::string &name)
{
  ProcessResult<int> result;

  if (internal_state == Net2State::STOPPED && name.empty() == false)
  {
    last_desire_name = name;
    this->name = name;

    if ( callbackGetServiceInfo )
    {
      ProcessResult<ppconsul::ServiceInfo> result2 = callbackGetServiceInfo(name);

      if (result2.success )
      {
         ppconsul::ServiceInfo info = result2.result;
         Start(info.address, info.port);
      }
    }
  }

  return result;
}

ProcessResult<int> Net2Subscriber::Start(const std::string &remote_ip, int remote_port)
{
  ProcessResult<int> result;
  thread_exit = false;

  if (remote_ip == "" || remote_port < 1000)
  {
    result.message = "Invalid arguments";
    result.success = false;
    return result;
  }

  if (internal_state != Net2State::STOPPED)
  {
    result.success = false;
    result.message = "Invalid state";
    return result;
  }

  try
  {
    reportLog("Starting subscriber " + getName(), LogType::INFO, section);

    this->remote_ip = remote_ip;
    this->remote_port = remote_port;

    path = "tcp://" + remote_ip + ":" + std::to_string(remote_port);

    socket = zmq_socket(zmqcontext,ZMQ_SUB);

    setState(Net2State::STARTED);

    //srand (time(NULL));
    //int v1 = rand() % 1000 + 1;

    //this->monitor_path = "inproc://sub" + name + ".inproc" + std::to_string(remote_port) + std::to_string(v1);

    thread_exit = false;

    //int rc = zmq_socket_monitor(socket,this->monitor_path.c_str(),ZMQ_EVENT_ALL);

    //monitor_thread = new boost::thread(&Net2Publisher::monitorThread,this);

    zmq_connect(socket,path.c_str());

    const char *filter = "";

    zmq_setsockopt(socket,ZMQ_SUBSCRIBE, filter, strlen (filter));

    socket_thread = new std::thread(&Net2Subscriber::socketThread,this);

    result.message = "Done";
    result.success = true;

    reportLog("Subscriber " + getName() + " is setup for  " + path, LogType::INFO, section);
  }
  catch (const std::runtime_error &e)
  {
    result.message = e.what();
    result.result = false;
    reportLog(e.what(), LogType::ERROR, section);
  }

  return result;
}
}
}
}
